ship_functions.py

```diff
@@ -0,0 +1,212 @@
+
+
+def weights_selection(proj_dir, choice):
+    '''Manage the weights file with returning the path after downloading file if
+    necessary'''
+        # Select weights file to load
+    if choice.lower() == "coco":
+        weights_path = proj_dir + "coco_trained_weights.h5"
+        # Download weights file
+        if not os.path.exists(weights_path):
+            utils.download_trained_weights(weights_path)
+    elif choice.lower() == "last":
+        # Find last trained weights
+        weights_path = model.find_last()
+    elif choice.lower() == "imagenet":
+        # Start from ImageNet trained weights
+        weights_path = model.get_imagenet_weights()
+    else:
+        weights_path = choice
+
+    return weights_path
+
+
+def train_configuration(logs_path, weights_path, weights):
+    '''Configure the MRCNN algorithme for the training'''
+    config = ship_config.ShipConfig()
+    config.display()
+
+    ship_model = model.MaskRCNN(mode="training", config=config,
+                                model_dir=logs_path)
+
+    #Exclude the last COCO layers because they require a matching number of classes
+    if weights.lower() == "coco":
+        ship_model.load_weights(weights_path, by_name=True, exclude=[
+            "mrcnn_class_logits", "mrcnn_bbox_fc",
+            "mrcnn_bbox", "mrcnn_mask"])
+    else:
+        ship_model.load_weights(weights_path, by_name=True)
+
+    return ship_model, config
+
+
+def launch_training(dataset_dir, train, val, local_model, config):
+    '''Launch Mask RCNN training'''
+    # Training dataset.
+    dataset_train = ship_dataset.ShipDataset()
+    dataset_train.load_ship(dataset_dir + "train/", train)
+    dataset_train.prepare()
+
+    # Validation dataset
+    dataset_val = ship_dataset.ShipDataset()
+    dataset_val.load_ship(dataset_dir + "val/", val)
+    dataset_val.prepare()
+
+    # Beginning of the training
+    print("Training network heads")
+    local_model.train(dataset_train, dataset_val,
+                      learning_rate=config.LEARNING_RATE,
+                      epochs=110,
+                      layers='heads')
+
+
+def detect_configuration(result_folder, weights_path):
+    '''Configure the MRCNN algorithme for the training'''
+    config = ship_config.ShipConfig()
+    config.display()
+
+    ship_model = model.MaskRCNN(mode="inference", config=config,
+                                model_dir=result_folder)
+
+    ship_model.load_weights(weights_path, by_name=True)
+
+    return ship_model, config
+
+
+def mask_analyse(mask):
+    '''Analyse the mrcnn mask result and return a string if exist'''
+    # We're treating all instances as one, so collapse the mask into one layer
+    mask = (np.sum(mask, -1, keepdims=True) >= 1)
+
+    if True in mask:
+        print("[INFO]: Ship(s) detected")
+        pixels = mask.T.flatten()
+        pixels = np.concatenate([[0], pixels, [0]])
+        ships = np.where(pixels[1:] != pixels[:-1])[0] + 1
+        ships[1::2] -= ships[::2]
```

Comment:
Can you add a comment about what's happening here? There is not enough context to understand it as an external reviewer

Reply:
@Ngoguey42 , @Toz3wm 
I would love to have your point of view of the changes brought. I left the old version in comment. Please, tell me if it is better or worst
```py
    if mask.any():
        logging.info(" Ship(s) detected")
        print("mask:", type(mask), mask.shape, mask)
        pixels = mask.T.flatten()

#        pixels = np.r_[0, pixels, 0]
#        ships = np.where(pixels[1:] != pixels[:-1])[0] + 1
        # Calculate the length and overwrite the result on the "end" location
#        ships[1::2] -= ships[::2]

        # Get an array with the indexs of biginnings and ends of 1
        ships = np.nonzero(np.ediff1d(pixels, to_end=0, to_begin=0))[0] + 1
        lengths = np.diff(np.hsplit(ships, 2))
        starts = np.delete(np.hsplit(ships, 2), 1, axis=1)
        # Get an array as [index_beginning length index_beginning length ...]
        ships = np.concatenate((starts, lengths), axis=1).flatten()

        return ' '.join(str(x) for x in ships)
    else:
        return None
```